flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoStats.java

```diff
@@ -44,7 +45,7 @@
     private final long endTime;
 
     /** Map of thread info samples by execution ID. */
-    private final Map<ExecutionAttemptID, List<ThreadInfoSample>> samplesBySubtask;
+    private final Map<Set<ExecutionAttemptID>, List<ThreadInfoSample>> samplesBySubtask;
```

Comment:
Does it make sense to retain the original key type? E.g. you'd flatten the set with `putAll`  into multiple entries. `Collection` as a key is prone to surprises.
You can then probably retain the original `matchExecutionsWithGateways` implementation (not sure though).

Reply:
Hmm... I am not sure about that:

1. There is a Set-to-List relationship. So we'd need to know which element from the list "belongs" to which element in the set. That means passing down not just Thread IDs, but the (ExecutionAttemptID,ThreadID) bundles all the way down to the MXBeans JVM sampler, plus preserving this relation in the result (`ThreadInfoSample` will have to contain the ExecutionAttemptID).
2. The changes in `matchExecutionsWithGateways` are sort of the core of the optimization, because this is where "bundling" requests by their `TaskManager` actually happens. 

What are the concerns with the defaults Set's hashCode?
- Is it the concern that elements could be added into the collection after it was "hashed" into a Map?
=> We could wrap it in a `Collections.unmodifiableSet(set)`
- Is it the default implementation that would potentially yield the same results for the "anagrams"?
[AbstractSet.java](https://www.cc.gatech.edu/computing/pag/tmp/html_dir/java/util/AbstractSet.java.html)
=> In that case we could implement a custom key type and go full Joshua Bloch on it (Chapter 11 in the 3rd edition and add multiplication with the previous "result" for each element while iterating)
- Is it the performance of recomputing the hashCode?
=> We could cache it
- Is there anything else I missed in terms of surprises?